### PR TITLE
Results and notebooks for benchmarks using hashing vectorizer (cuML+Dask vs Apache Spark)

### PR DESCRIPTION
**Pull request by akaanirban**
This PR adds 3 note books and benchmarking results between scikit-learn (baseline), Apache Spark and cuML+Dask pipelines for NLP data processing.

### CODE REVIEW

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb [hunk 1]
```diff
@@ -0,0 +1,1650 @@
+{
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #38.                with SimpleTimer() as overalltimer:
</pre><p>We should put this function in a function to make this easy to follow. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #40.                        data = dask_cudf.read_parquet(parquet_path,storage_options={'anon': True, 'use_ssl': False}).sample(frac=data_frac, random_state=seed)
</pre><p>Remove extra arguments being used here as well as remove data_fra. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

File: tfidf-benchmark/NLP-pipeline-sklearn.ipynb [hunk 2]
```diff
@@ -0,0 +1,536 @@
+{
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-sklearn.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def performance_numbers(parquet_path, worker_counts=[1], samples=1, data_frac=0.2, result_path="./results.pkl", 
</pre><p>Remove <code>data_frac</code> as that's no longer used to ensure we have similar results. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-sklearn.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #15.                    with SimpleTimer() as timer:
</pre><p>We should put the overall workflow that follows in a function so that it's easier to follow. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: tfidf-benchmark/NLP-pipeline-spark.ipynb [hunk 3]
```diff
@@ -0,0 +1,673 @@
+{
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-spark.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #16.                        data = spark.read.parquet(parquet_path).sample(withReplacement=False, fraction=data_frac, seed=seed)
</pre><p>We need to persist this too to benchmark read_parquet correctly. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-spark.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #36.                        featurizedData.head(0)
</pre><p>Same here. I suggest just benchmarking till this bit with head(0) . </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-spark.ipynb:1]: <p>Yup makes sense. </p>

Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-spark.ipynb:1]: <p>You mean removing the IDF part? I can persist the <code>featurizedData</code>  so that it performs a compute there. But I do not think in a tf-idf pipeline it makes sense to stop at <code>hashingTF</code>  . The latencies of different parts of the pipeline are compartmentalized anyway. We can choose not to use the <code>hashingTF</code>  one even though it is in the complete pipeline. </p>

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb (hunk 1 above)
Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <p>I think <code>data_frac</code>  if fine as is. For completeness one may try to run this somewhere else with a smaller chunk of dataset. However I should remove the <code>model_args={}, model_kwargs={}</code>  though. Good Catch.</p>

**Comment by akaanirban** [comment]
To Check: 
Why cuML + Dask is slower without intermediate persists/compute_chunk_sizes ?

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb [hunk 4]
```diff
@@ -0,0 +1,4769 @@
...
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #50.        print(hashing_vectorized.shape)
</pre><p>You are computing shape at end of every step. This is the reason for 2x the time. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Thanks fo the work on this .  

We still need to make the following changes: 

1. Remove the binary files  (.pkl files)
2. Flesh out the utility functions  into a util.py file to make the notebook more readaale.

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb (hunk 4 above)
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    import nltk
</pre><p>Remove unused imports</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #29.                             .str.replace(PUNCTUATIONS, [" "]*len(PUNCTUATIONS), regex=False)
</pre><p> </p><p>Why does  	<code>[" "]</code>  have to be a list ? </p><p>Also, <code>replace_tokens</code>   might be faster here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #41.        vectorizer = HashingVectorizer(stop_words='english', preprocessor=None)
</pre><p>You are replacing stop_words twice. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #83.        print(f"Workers:{n}, Sample Run:{i}, Finished loading data in {sample_record['data_read']}s")
</pre><p>The time should be seconds instead on ns to make ti more readable. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: tfidf-benchmark/NLP-pipeline-spark.ipynb [hunk 5]
```diff
@@ -0,0 +1,1247 @@
...
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-spark.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #16.        if persist_flag:
</pre><p>this should be <code>if </code> benchmark now . </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb (hunk 4 above)
Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: Removed this. However while it is not 2x anymore, the latency is still larger.

Comment (VibhuJawa) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <p>Can you share what the errors that you see <code>intermittently</code>  ? </p>

Comment (VibhuJawa) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <p>Lets pair on this tomorrow. Something seems to be wrong with the large time. </p>

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb [hunk 6]
```diff
@@ -0,0 +1,2955 @@
...
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    from utils import SimpleTimer, ResultsLogger, scale_workers
Line #11.  from utils import visualize_data_cuml as visualize_data
</pre><p> </p><p>Import these in a different cell explaining with a comment explaining that you are importing this for benchmarking work.    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">&nbsp; &nbsp; if benchmark:
&nbsp; &nbsp; &nbsp; &nbsp; with dask.annotate(workers=set(client.has_what().keys())):
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; result = client.persist(result)
&nbsp; &nbsp; &nbsp; &nbsp; wait(result)
</pre><p> </p><p>We should add these four lines to a function as this is repeated across the workflow. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (VibhuJawa) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: Another question ? 

```
 with dask.annotate(workers=set(client.has_what().keys())):
 ```
 
 Is this line needed as to me it seems that it will scale to all the workers that client has ?

Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: 1. That makes sense. Would you prefer if I just did `client.persist()` an then `wait` instead of the dask annotations?
2. Actually It would not since I am scaling the number of workers to a fixed number before each run starts.

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb [hunk 7]
```diff
@@ -0,0 +1,1874 @@
...
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #51.    def tfidf_transformer_POC(data, client):
</pre><p>This function is not used. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    groupby
</pre><p>unnamed:0 should be removed . </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: tfidf-benchmark/NLP-pipeline-sklearn.ipynb [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
```
Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-sklearn.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    data
</pre><p>Remove Unnamed:0 column from here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (VibhuJawa) [inline tfidf-benchmark/NLP-pipeline-sklearn.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    groupy
</pre><p>remoce unamed:0 column</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: tfidf-benchmark/Visualize.ipynb [hunk 9]
```diff
@@ -0,0 +1,1558 @@
+{
```
Comment (VibhuJawa) [inline tfidf-benchmark/Visualize.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    groupy_sklearn
</pre><p>Remove unmaed:0 column (index in all probability) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (VibhuJawa) [inline tfidf-benchmark/Visualize.ipynb:1]: <p>Love this graph. Could you add cuml+Dask+n_gpus to the legend</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/rapids-examples/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: tfidf-benchmark/NLP-pipeline-cuML.ipynb (hunk 7 above)
Comment (akaanirban) [reply tfidf-benchmark/NLP-pipeline-cuML.ipynb:1]: Removed.

**Review by VibhuJawa — APPROVED** [review]
LGTM, Thanks a ton for the work on this @akaanirban . The code will serve as good example for people trying to use RAPIDS for NLP as well as a good standard to follow for benchmarking.
